### PR TITLE
MergeTree: Move Obliterate to use local references

### PR DESCRIPTION
**Pull request by anthony-murphy**
This change was developed in a topic branch by me, @jzaffiro, and @titrindl, as such the changes have been incrementally reviewed. 

The overall change moves the implementation of merge-trees obliterate operation from a segment-based approach to a local reference-based approach. This motivation behind this change is to support further changes to the obliterate operation such that it can have different range expansion properties, or sidedness. The changes are much more easily supported by the local reference-based approach as it allows us to maintain and evaluate ranges independently of the physical segment layout in the tree. Specifically, this will allow the endpoints of the obliterate operation to exist beyond the segments that are actually obliterated, so that the range will naturally encompass changes between that endpoint, and the obliterated segments, which was hard to compute in the existing implementation.

### CODE REVIEW

File: packages/dds/merge-tree/src/mergeTree.ts [hunk 1]
```diff
@@ -1964,29 +1901,65 @@ export class MergeTree {
...
-			if (startSide) segment.startSide = startSide;
-			if (endSide) segment.endSide = endSide;
-
 			if (
 				clientId !== segment.clientId &&
 				segment.seq !== undefined &&
 				seq !== UnassignedSequenceNumber &&
 				(refSeq < segment.seq || segment.seq === UnassignedSequenceNumber)
 			) {
+				// why do we need this? we don't need it for the insert + remove case?
```
[the remark this answers is not in the document]
Comment (titrindl) [reply packages/dds/merge-tree/src/mergeTree.ts:1960]: We need this in order to calculate partial lengths for the obliterating client in between the move op's refSeq and the greater of the move op's or the insert op's seq. The obliterator made the move op before it was aware of this concurrent insert, so it should never include the inserted segment's length. Other clients, especially the inserter, might include the inserted segment's length for refSeqs up to the seq of the obliterate.

File: packages/dds/merge-tree/src/mergeTree.ts [hunk 2]
```diff
@@ -1964,29 +1901,65 @@ export class MergeTree {
...
+				// there are subtle differences in the visibility of the client doing to remove
+				// however, i'm not aware of invariants in the code that would break due to that
```
Comment (anthony-murphy) [inline packages/dds/merge-tree/src/mergeTree.ts:1962]:
```suggestion
```

File: packages/dds/merge-tree/src/mergeTree.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (anthony-murphy) [reply packages/dds/merge-tree/src/mergeTree.ts]: Thanks! this was just left over from when i was starting to investigate. nothing to be done.

**Review by titrindl — APPROVED** [review]

File: packages/dds/merge-tree/api-report/merge-tree.legacy.alpha.api.md [hunk 3]
```diff
@@ -641,6 +641,24 @@ export interface MergeTreeRevertibleDriver {
     removeRange(start: number, end: number): void;
 }
 
+// @alpha (undocumented)
+export interface ObliterateInfo {
```
Comment (Abe27342) [inline packages/dds/merge-tree/api-report/merge-tree.legacy.alpha.api.md:645]: SegmentGroup is deprecated, so can we deprecate this as well to discourage new usage?

Comment (anthony-murphy) [reply packages/dds/merge-tree/api-report/merge-tree.legacy.alpha.api.md:645]: i had the same thought, but decided against it as the suppressions are a bit annoying, but will make the change, as it feels a bit safer even if it is odd to add a type that is directly deprecated :)

**Review by Abe27342 — APPROVED** [review]

**Pull request merged**